### PR TITLE
Feature/3 migrations and start to implement jwt tokens

### PR DESCRIPTION
**Pull request by VitaliKhileuski**

### CODE REVIEW

File: HotelReservation/Api/Program.cs [hunk 1]
```diff
@@ -1,26 +1,35 @@
...
             Host.CreateDefaultBuilder(args)
                 .ConfigureWebHostDefaults(webBuilder =>
                 {
-                    webBuilder.UseStartup<Startup>();
+                    webBuilder.UseStartup<Startup>().UseSerilog();
                 });
+
+        public static void ConfigureLogger()
+        {
+            Log.Logger = new LoggerConfiguration().WriteTo.File(@"log.txt").CreateLogger();
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Api/Program.cs]: extract logger configuration to appsettings

File: HotelReservation/Api/Startup.cs [hunk 2]
```diff
@@ -27,14 +28,37 @@ public Startup(IConfiguration configuration)
...
+                    x => x.MigrationsAssembly("Api")));
+            services.AddScoped<HashPassword>();
+            services.AddScoped<ITokenService, TokenService>();
+            services.AddScoped<IAuthenticationService, AuthenticationService>();
             services.AddControllers();
+            services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
+                .AddJwtBearer(options =>
+            {
+                options.RequireHttpsMetadata = false;
+                options.TokenValidationParameters = new TokenValidationParameters
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Api/Startup.cs:55]: extract TokenValidationParameters to appsettings

File: HotelReservation/Api/Startup.cs [hunk 3]
```diff
@@ -27,14 +28,37 @@ public Startup(IConfiguration configuration)
...
+                    ClockSkew = TimeSpan.Zero
+                };
+            });
         }
+        
 
         // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
         public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
         {
+
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Api/Startup.cs]: remove extra line

File: HotelReservation/Api/Startup.cs [hunk 4]
```diff
@@ -45,11 +69,18 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
             app.UseRouting();
 
             app.UseAuthorization();
+            app.UseAuthentication();
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Api/Startup.cs:90]: app.UseAuthorization();
app.UseAuthentication();
is it the right order?

File: HotelReservation/Api/Startup.cs [hunk 5]
```diff
@@ -45,11 +69,18 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
 
             app.UseEndpoints(endpoints =>
             {
                 endpoints.MapControllers();
             });
+            using var serviceScope = app.ApplicationServices
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Api/Startup.cs:102]: extract to separate method

File: HotelReservation/Api/appsettings.json [hunk 6]
```diff
@@ -6,5 +6,18 @@
       "Microsoft.Hosting.Lifetime": "Information"
     }
   },
-  "AllowedHosts": "*"
-}
+  "AllowedHosts": "*",
+  "Salt": "T_CpC.rqbvX65ycC!dhK4I-0G(QAzSId",
+
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Api/appsettings.json]: extra line

File: HotelReservation/Business/DTO/LoginUserDTO.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Business.DTO
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Business/DTO/LoginUserDTO.cs]: DTO is a pattern (abstract name) -> rename to smth meaningfull

File: HotelReservation/Business/DTO/LoginUserDTO.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+{
+    public class LoginUserDTO
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Business/DTO/LoginUserDTO.cs]: replace DTO

File: HotelReservation/Business/Interfaces/IAuthenticationService.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Business.DTO;
+
+namespace Business.Interfaces
+{
+    public interface IAuthenticationService
+    {
+        Task<string> Login(LoginUserDTO user);
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Business/Interfaces/IAuthenticationService.cs]: dto...

File: HotelReservation/Business/Services/AuthenticationService.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+using HotelReservation.Data.Entities;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Configuration;
+using Org.BouncyCastle.Crypto.Tls;
+
+namespace Business.Services
+{
+    public class AuthenticationService : IAuthenticationService
+    {
+
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Business/Services/AuthenticationService.cs:23]: extra line

File: HotelReservation/Business/Services/AuthenticationService.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+using System.Diagnostics.Contracts;
+using System.Text;
+using System.Threading.Tasks;
+using Business.DTO;
+using Business.Interfaces;
+using HotelReservation.Data;
+using HotelReservation.Data.Entities;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Configuration;
+using Org.BouncyCastle.Crypto.Tls;
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Business/Services/AuthenticationService.cs]: dont use any references not from provided by Microsoft at this moment

File: HotelReservation/Data/Configurations/RoomEntityConfiguration.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using HotelReservation.Data.Entities;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+namespace HotelReservation.Data.Configurations
+{
+    class RoomEntityConfiguration : IEntityTypeConfiguration<RoomEntity>
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Data/Configurations/RoomEntityConfiguration.cs]: access modifier is missed

File: HotelReservation/Data/Interfaces/IRepository.cs [hunk 13]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace HotelReservation.Data.Interfaces
+{
+   public interface IRepository<T> where T : class
```
Comment (DzmitryAliakseichyk) [inline HotelReservation/Data/Interfaces/IRepository.cs:6]: Do u really need so fat interface?
sol **I**d

**Review by DzmitryAliakseichyk — CHANGES_REQUESTED** [review]
- web api should use /api/
 - I don't want to see "dto" suffix rename it in layers to smth useful
 - remove extra lines
 - class access modifiers are missed
 - extract base repository (do not copy/paste)

File: HotelReservation/Api/Program.cs (hunk 1 above)
Comment (VitaliKhileuski) [reply HotelReservation/Api/Program.cs]: extracted

File: HotelReservation/Api/Startup.cs (hunk 2 above)
Comment (VitaliKhileuski) [reply HotelReservation/Api/Startup.cs:55]: extracted

File: HotelReservation/Api/Startup.cs (hunk 3 above)
Comment (VitaliKhileuski) [reply HotelReservation/Api/Startup.cs]: removed

File: HotelReservation/Api/Startup.cs (hunk 4 above)
Comment (VitaliKhileuski) [reply HotelReservation/Api/Startup.cs:90]: my bad, already replaced.

File: HotelReservation/Api/Startup.cs (hunk 5 above)
Comment (VitaliKhileuski) [reply HotelReservation/Api/Startup.cs:102]: extracted

File: HotelReservation/Api/appsettings.json (hunk 6 above)
Comment (VitaliKhileuski) [reply HotelReservation/Api/appsettings.json]: deleted

File: HotelReservation/Business/DTO/LoginUserDTO.cs (hunk 7 above)
Comment (VitaliKhileuski) [reply HotelReservation/Business/DTO/LoginUserDTO.cs]: renamed

File: HotelReservation/Business/Services/AuthenticationService.cs (hunk 10 above)
Comment (VitaliKhileuski) [reply HotelReservation/Business/Services/AuthenticationService.cs:23]: removed

File: HotelReservation/Business/Services/AuthenticationService.cs (hunk 11 above)
Comment (VitaliKhileuski) [reply HotelReservation/Business/Services/AuthenticationService.cs]: already deleted

File: HotelReservation/Data/Configurations/RoomEntityConfiguration.cs (hunk 12 above)
Comment (VitaliKhileuski) [reply HotelReservation/Data/Configurations/RoomEntityConfiguration.cs]: added

File: HotelReservation/Data/Interfaces/IRepository.cs (hunk 13 above)
Comment (VitaliKhileuski) [reply HotelReservation/Data/Interfaces/IRepository.cs:6]: divided it into two interfaces

**Pull request merged**